### PR TITLE
add network operating models

### PR DESCRIPTION
**Pull request by deitch**
As discussed last week in person, capture the networking _models_ for Edge Container workloads, and the networking _implementations_ that support those models.

cc @eriknordmark @rvs 

Please add any corrections.

### CODE REVIEW

File: docs/NETWORK-MODELS.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
+# Network Models
+
+What are the network models that EVE supports for Edge Container workloads, and how are they implemented?
```
Comment (rvs) [inline docs/NETWORK-MODELS.md]: I am not sure we need to keep repeating "Edge Container workloads" through this document. Why not stick to simply calling those "Edge Containers"?

File: docs/NETWORK-MODELS.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+The first half of this document describes _what_ those network models are and how they impact workload communications. The second half describes _how_ those models are implemented.
+
+## Models
+
+Fundamentally, each Edge Container workload needs to communicate with services outside itself. Those can be one or more of:
+
+* other Edge Container workloads on the same device
+* the network to which the device is attached
```
Comment (rvs) [inline docs/NETWORK-MODELS.md]: We actually don't have a notion of "the network to which the device is attached"

File: docs/NETWORK-MODELS.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+* a network (VPC) in the public cloud
+* no networking at all
+
+It is possible to define all network requirements using communications permissions model. For example:
+
+* workload-to-workload: mesh (between devices) or switch (same device)
+* workload-to-cloud: VPN
+* workload-to-host-network-NAT: switche
+* workload-to-host-network-transparent: bridge
+
```
Comment (rvs) [inline docs/NETWORK-MODELS.md]: Again, we really shouldn't be confusing our readers with anything that has to do with the host. Host is completely invisible to edge containers.

File: docs/NETWORK-MODELS.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+In principle, a workload definition does not define any of its methodologies, networks or technologies. It simply defines what it should be able to connect to. The Controller then is responsible for translating those networking permissions to implementations passed on to EVE.
+
+## Implementations
+
+EVE currently supports four primary networking implementations - in addition to "no network" - to support the above models. These are:
+
+* Bridged - EVE sets up a transparent L2 bridge between the workloads and the underlying device network. The responsibility for IP address management (IPAM) is handed off to the underlying fabric.
```
Comment (rvs) [inline docs/NETWORK-MODELS.md]: Again -- there's no underlying device network.

File: docs/NETWORK-MODELS.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+* Switched - EVE sets up a switch between the workloads and the underlying device network. EVE provides IPAM to the workloads, and NAT between the workloads and the device network.
```
Comment (rvs) [inline docs/NETWORK-MODELS.md]: It is called Local

File: docs/NETWORK-MODELS.md (hunk 1 above)
Comment (deitch) [reply docs/NETWORK-MODELS.md]: I am good either way. I felt I was under-using the term Edge Containers, so I started using them in a loose hybrid fashion. I will call them Edge Containers throughout.

File: docs/NETWORK-MODELS.md (hunk 2 above)
Comment (deitch) [reply docs/NETWORK-MODELS.md]: But don't we have the "bridged" option, where it just is connected to an underlying network?

File: docs/NETWORK-MODELS.md [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+* Edge Container-to-host-network-NAT: switche
+* Edge Container-to-host-network-transparent: bridge
+
+In principle, a Edge Container definition does not define any of its methodologies, networks or technologies. It simply defines what it should be able to connect to. The Controller then is responsible for translating those networking permissions to implementations passed on to EVE.
+
+## Implementations
+
+EVE currently supports four primary networking implementations - in addition to "no network" - to support the above models. These are:
+
+* Bridged - EVE sets up a transparent L2 bridge between the Edge Containers and the underlying device network. The responsibility for IP address management (IPAM) is handed off to the underlying fabric.
```
Comment (eriknordmark) [inline docs/NETWORK-MODELS.md]: Called "switch"

File: docs/NETWORK-MODELS.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+* Switched - EVE sets up a switch between the Edge Containers and the underlying device network. EVE provides IPAM to the Edge Containers, and NAT between the Edge Containers and the device network.
```
Comment (eriknordmark) [inline docs/NETWORK-MODELS.md]: Doesn't exist.

There is a "local" which includes IPAM (plus DNS etc) where the NAT to the outside world is optional. The NAT is selected by specifying an adapter (XXX have to check that the term is in the API for the network instance)

File: docs/NETWORK-MODELS.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+* Mesh - EVE sets up a [lisp](http://lispers.net) mesh network between Edge Containers, whether on the same device or on multiple devices.
+* Cloud - EVE sets up a VPN connection between Edge Containers and a VPN endpoint in a public cloud VPC.
+
+It is possible for one device to have multiple Edge Containers, each using a different network implementation.
```
Comment (eriknordmark) [inline docs/NETWORK-MODELS.md]: I wouldn't call it "network implementation" but instead call it "network instances".
Furthermore, an Edge Container can be attached to more than one network instance.

**Comment by deitch** [comment]
I redid the doc (almost) from scratch. Please review again.

File: docs/NETWORK-MODELS.md [hunk 9]
```diff
@@ -0,0 +1,87 @@
+# Network Models
+
+What are the network models that EVE supports for Edge Container workloads, and how are they implemented?
+
+The first half of this document describes _what_ those network models are and how they impact Edge Container communications. The second half describes _how_ those models are implemented.
+
+## Models
+
+Fundamentally, each Edge Container may need to communicate with services outside itself. Those can be one or more of:
+
```
Comment (rvs) [inline docs/NETWORK-MODELS.md]: Small nit: not sure that 'Those can be one or more of'... no networking at all reads that well.

File: docs/NETWORK-MODELS.md [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+In principle, an Edge Container definition does not define any of its methodologies, networks or technologies. It simply defines what it should be able to connect to. The Controller then is responsible for translating those networking permissions to implementations passed on to EVE.
+
+Put in other terms, the various complex methods of networking are the responsibility of the _Controller_ to define and expose to users. EVE itself, on the other hand, is responsible solely for providing basic networking building blocks that enable all of the necessary forms of communication, and which a Controller can meld together to provide the necessary networking services.
+
+## Definitions
+
+* `network`: a joining together of one or more endpoints that will be able to communicate with each other via packets
+* `endpoint`: a unique connection point to a network, includes Edge Containers and one or more communications media
+* `Edge Container`: a single running workload on a single device, which can be connected as an endpoint to 0, 1 or multiple networks simultaneously
```
Comment (rvs) [inline docs/NETWORK-MODELS.md]: Wouldn't it clearer if the Edge Container and port were a sub-bullets of endpoint?

File: docs/NETWORK-MODELS.md [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+## Responsibilities
+
+From the perspective of EVE on a single device, which is its entire scope of responsibility, EVE needs to be informed of the following:
+
+* that a network should exist on the device, and a representative name for it
+* what the network type is: L2 or L3 (see below)
+* the endpoints that are to be connected to that network (see below)
+
+Upon receiving these, EVE should either create a network that matches the specifications and connect endpoints, or update one if it already exists.
+
```
Comment (rvs) [inline docs/NETWORK-MODELS.md]: Should we make it clear that the mechanism of "receiving" is via generic config?

File: docs/NETWORK-MODELS.md (hunk 9 above)
Comment (deitch) [reply docs/NETWORK-MODELS.md]: No problem. Emphasized the "may" and removed the "no networking at all".

File: docs/NETWORK-MODELS.md (hunk 10 above)
Comment (deitch) [reply docs/NETWORK-MODELS.md]: I don't think so for this one. It is just a list of definitions. I tried it both ways, find it easier this way.

File: docs/NETWORK-MODELS.md (hunk 11 above)
Comment (deitch) [reply docs/NETWORK-MODELS.md]: Yes, no problem. Updating.

**Comment by deitch** [comment]
I am trying to understand the points you are highlighting @eriknordmark but struggling with it. It appears there are two points you are making:

* ports, specifically serial ports I the context of networking
* the number and types of networks

I will address each in turn.

### ports

My understanding is that, from a network perspective - i.e. a network that runs on an EVE device - there are only a few kinds of endpoints that EVE will connect to it:

* an EdgeContainer - I think we all are in agreed understanding here
* a port, which is backed by some physical adapter, which could be a NIC, or an LTE card (which really is just a NIC), or a serial port over which we can run some L2 connectivity to create something that can transit packets, e.g. PPP
* a VPN connection

Is that incorrect? I had understood (based on the Slack discussion) that a `port` is nothing more or less than a physical adapter (NIC, serial, whatever) that can be connected to a network running on the device.

### network types

I get that from the _entire system_'s perspective (multiple EVEs managed by a Controller) that there are 4 network types, as described above: switched (L2), local (L3+NAT to any ports), cloud (L3+VPN to cloud endpoint), mesh (L3+VPN to other devices).

But my understanding was that from  _a single EVE device_'s perspective, there really are just two kinds of networks:

* L2 (switched)
* L3 (local)

Either can attach a port endpoint. With an L2, it is just connected; with an L3, it is given a NAT layer.

Additionally, L3 can also attach a VPN endpoint to somewhere (anywhere; the local EVE device doesn't really care much where the far end of the VPN is). The VPN endpoint doesn't have NAT in front of it; just the ports (which represent local physically attached adapters).

Those are the only building blocks that EVE understands. The Controller then builds additional network types by:

* Cloud: L3/Local on device+VPN endpoint whose remote end is in the cloud
* Mesh: L3/Local on device+VPN endpoints whose remote ends are other devices

Where is that incorrect? How much more knowledge does EVE have (or, more importantly, need to have)?

The _whole system_ then constructs

**Comment by eriknordmark** [comment]
Ports is a networking concept; adapters is an I/O concept. For this to match the description they need to be described separately (under an umbrella name of "workload connectivity" which consists of "workload networking" and "workload I/O (connectivity)".

> a port, which is backed by some physical adapter, which could be a NIC, or an LTE card (which really is just a NIC), or a serial port over which we can run some L2 connectivity to create something that can transit packets, e.g. PPP

No.
What EVE supports is assigning serial ports (and audio, USB, FPGA devices) to a workload. That is part of connectivity but has nothing to do with networking. The workload kernel runs the serial port(s) i.e., /dev/ttySx
While there might be some LTE modem involving a serial port or USB connectivity or whatever, that is not something we expose to the workload. The workload can get network connectivity (a virtual Ethernet NIC) which happens to be connected to the outside world using the LTE modem, but that is irrelevant for the assignment.

> a VPN connection
No.

If we limit the discussion to network connectivity (hence serial ports, USB, etc is out of the picture), then a network in EVE can be attached to zero or more external ports, and provide different service over those ports. The type of service is a function of the type of network as follows:
switch: L2 service. Can have zero or one ports attached to the network.
local: NAT out the port(s) (can have more than one port attached to the network)
vpn: The IPsec VPN uses the port to set up the tunnels.
mesh: LISP can use one or more ports for the overlay.
Note that NAT is only in 1 out of 4.

> Is that incorrect? I had understood (based on the Slack discussion) that a port is nothing more or less than a physical adapter (NIC, serial, whatever) that can be connected to a network running on the device.

That description would be confusing, because we're likely to have another layer of indirection to using VPN and LAG to create ports to be used for IP connectivity (hence be used for local, mesh, and cloud networks), where there is not a 1-1 mapping between the NIC/RJ-45 and the port used by the network. 

> I get that from the entire system's perspective (multiple EVEs managed by a Controller) that there are 4 network types, as described above: switched (L2), local (L3+NAT to any ports), cloud (L3+VPN to cloud endpoint), mesh (L3+VPN to other devices).

Incorrect. The 4 types are realized in EVE.
The aspect that there are remote participants (for any of the 4 types) is completely out of scope for this discussion.

> But my understanding was that from a single EVE device's perspective, there really are just two kinds of networks:
Incorrect.

> Either can attach a port endpoint. With an L2, it is just connected; with an L3, it is given a NAT layer.
Incorrect.
(I think I already stated the two most recent points last night; I'm starting to feel like a broken record.)


In particular, for local networks EVE provisions the NAT (out one or more ports) and runs DHCP plus DNS to provide the workload with connectivity.
For cloud networks EVE is handed the VPN configuration from the controller, sets up StrongSWAN IPsec with that configuration, hands out static IP addresses (which it got from the controller) to the workloads using DHCP, and runs a DNS proxy. No NAT in that case.
For mesh networks we can use IPv4 or IPv6, with the difference being that for IPv6 we form the addresses for the workloads using a hash of the public key, whereas for IPv4 the controller passes static IP addresses for the workload. In either case EVE provisions the LISP overlay using the LISP configuration it receives from the controller. And runs DHCP plus DNS. No NAT in this case.

Hence EVE is very much involved in realizing all 4 forms of networks.

But high-order question is whether this document should be about connectivity (and include serial, USB, etc assignment) or be about workload network connectivity (and exclude all mention of serial, adapters, etc)

**Comment by eriknordmark** [comment]
I think it would make sense to add a separate document about Workload Connectivity, which can reference this Workload Networking document, and also reference a separate TBD Workload I/O - Adapter assignment

**Comment by deitch** [comment]
A lot here, I will try to deduplicate, but might fall afoul.

> But high-order question is whether this document should be about connectivity (and include serial, USB, etc assignment) or be about workload network connectivity (and exclude all mention of serial, adapters, etc)

Workload network connectivity. I am an EC, I see a NIC (or 2 or 10) provided by EVE. Other connectivity is out of scope and will be for a separate doc sometime later. Plenty to chew on here. The fact that we can take a physical device (NIC, serial port, entire PCI bus, anything at all) and assign it directly to an EC is out of scope of this document.

> I think it would make sense to add a separate document about Workload Connectivity, which can reference this Workload Networking document, and also reference a separate TBD Workload I/O - Adapter assignment

Yup... once this is done.

> I'm starting to feel like a broken record

Which may indicate that this is too complex (or at least described so), if we have gone over this multiple times in Slack and here, and a bunch of knowledgeable people cannot describe it in a way that is clear for outsiders to grasp. But we keep trying!

> If we limit the discussion to network connectivity (hence serial ports, USB, etc is out of the picture)

Yes!

Going to try and separate this into categories.

### Port Definition

> a network in EVE can be attached to zero or more external ports, and provide different service over those ports. 

I am fine with that, but in the Slack discussion, it was made clear to me that a network (virtual bridge or virtual switch) is connected to 2 potential things: 

* EC (fine, we know that)
* "port", which is a virtual concept that is backed by some physical connectivity to off-device, i.e. external NIC or maybe a serial port with PPP or something else. 

Are you saying that we are actually at the historical definition of a port on a virtual network, i.e. not some virtual concept special to EVE, but is just a a NIC or other physical device that provides network connectivity off-device? And it is managed in dom0 (or wherever the virtual switch/bridge exists), and we just connect the port (actual NIC) to the virtual network?

How does that align with:

> we're likely to have another layer of indirection to using VPN and LAG to create ports to be used for IP connectivity (hence be used for local, mesh, and cloud networks)

This implies a "port" _is_ a virtual NIC on the virtual network that is backed by some connectivity. Are you saying, e.g. that a VPN endpoint is (or could also be) a "port", where EVE sets up the VPN connection over some real off-device NIC, then exposes the VPN endpoint to the virtual network as a "port" on the network? If so, then a "port" _can_ be backed by a physical NIC or a VPN or something else, it just looks like a network endpoint (node, switch, router, etc.). 

How would you define a "port" concisely enough to have someone understand it? 

> where there is not a 1-1 mapping between the NIC/RJ-45 and the port used by the network

In what scenario would there not be? Bonding? VPN? Something else?

### Network Types

> The 4 types are realized in EVE

That was what I had thought the first time around; the Slack discussion, however, indicated that there were just the two (L2 & L3), with the ability to add a VPN to L3. That was all EVE knew about. "Cloud" was just the Controller telling EVE to create an L3 with a VPN endpoint, which turned it into "cloud". Same for "mesh". It gave the Controller the ability to _create_ "cloud" or "mesh" by combining L3 with VPN, but that it was not something native to EVE itself. You clearly are saying that is not true?

> switch: L2 service. Can have zero or one ports attached to the network.

Just 0 or 1? New to me; happy to add it.

> local: NAT out the port(s) (can have more than one port attached to the network)

So it can have 0, 1 or N ports (i.e., as described above, physical devices capable of providing off-device network traffic), and each one has its own NAT?

> vpn: The IPsec VPN uses the port to set up the tunnels.

How is this different than above? If I say, "give me EC2 and EC3 connected over L3 with a VPN connection to the cloud", is this _not_ just L3/local with an IPSec VPN set up that connects using an assigned NIC ("assigned" = "assigned to the network")? Or is this implemented somehow differently fundamentally? Somehow we have to wire EC2 and EC3 to each other and to the VPN connection? Or do we actually say, "no, only one EC per VPN connection, so EC2 gets one connection, EC3 gets another, and if you want EC2 and EC3 to talk, add them to a separate L3/local network"?

> mesh: LISP can use one or more ports for the overlay.

So EVE _is_ aware of mesh? It treats it as a fundamentally different networking beast than L3 or L2 but with VPN? How does it actually implement it? You are saying it does it differently than "I will set up an L3 network and VPN connect to another device"?

> Note that NAT is only in 1 out of 4.

Yes. Only on local. Got it.

> In particular, for local networks EVE provisions the NAT (out one or more ports) and runs DHCP plus DNS to provide the workload with connectivity

I definitely got that earlier, but missed putting it in on the latest revision. Will add it.

> That description would be confusing

I am living it! 😂

> For cloud networks EVE is handed the VPN configuration from the controller

OK.

> sets up StrongSWAN IPsec with that configuration

Where? In the same domain as a virtual network (i.e. right now dom0)?

> hands out static IP addresses (which it got from the controller) to the workloads using DHCP, and runs a DNS proxy. No NAT in that case.

Which sounds a lot like an L3/local network that is different from the usual L3/local only in that it:

* does not use NAT
* hands out IPs it got from the controller rather than from a pool
* Does not run NAT to the VPN (what about if it has a NIC port assigned? Or is that not allowed?)

> For mesh networks we can use IPv4 or IPv6, with the difference being that for IPv6 we form the addresses for the workloads using a hash of the public key, whereas for IPv4 the controller passes static IP addresses for the workload. 

OK

> In either case EVE provisions the LISP overlay using the LISP configuration it receives from the controller. And runs DHCP plus DNS. 

So this is a completely distinct network from L3/local? No common behaviour at all? Or is it still a virtual bridge?

>No NAT in this case.

What if a NIC port is attached? Or is that not allowed here, and connectivity is only via the LISP network?

**Comment by deitch** [comment]
@eriknordmark, I rewrote this based on our conversation last week. Please take a fresh look and see if it captures it.

It need not be complete, but should not be incorrect.

**Comment by rvs** [comment]
I'd like to merge this. @nordmark please holler if you disagree.

**Pull request merged**